site/content/protocols/pickup/3.0/readme.md

```diff
@@ -0,0 +1,249 @@
+
+This protocol is the explicit companion to the implicit method of picking up messages.
+
+## Roles
+There are two roles in this protocol: 
+
+- `mediator`: The agent that has messages waiting for pickup by the `recipient`.
+- `recipient`: The agent who is picking up messages from the `mediator`.
+
+## Connectivity
+
+This protocol consists of three different message requests from the `recipient` that should be replied by the `mediator`:
+
+1. Status Request -> Status
+2. Delivery Request -> Message Delivery
+3. Message Received -> Status
+4. Live Mode  -> Status or Problem Report
+
+## States
+
+This protocol follows the request-response message exchange pattern, and only requires the simple state of waiting for a response or to produce a response. 
+
+Additionally, the `return_route` header extension must be set to `all` in all request submitted by the `recipient`.
+
+## Basic Walkthrough
+
+The `status-request` message is sent by the `recipient` to the `mediator` to query how many messages are pending.
+
+The `status` message is the response to `status-request` to communicate the state of the message queue.
+
+The `delivery-request` message is sent by the `recipient` to request delivery of pending messages.
+
+The `message-delivery` message is the response to the `delivery-request` to send queued messages back to the `recipient`.
+
+The `message-received` message is sent by the `recipient` to confirm receipt of delivered messages, prompting the `mediator` to clear messages from the queue.
+
+The `live-delivery-change` message is used to set the state of `live_delivery`.
+
+When _Live Mode_ is enabled, messages that arrive when an existing connection exists are delivered over the connection immediately, rather than being pushed to the queue. See _Live Mode_ below for more details.
+
+## Security
+
+This protocol expects messages to be encrypted during transmission, and repudiable. 
+
+## Composition
+
+Supported Goal Code | Notes
+--- | ---
+                     |       
+                     |       
+
+
+
+## Message Reference
+
+### Status Request
+Sent by the `recipient` to the `mediator` to request a `status` message.
+
+Message Type URI: `https://didcomm.org/messagepickup/3.0/status-request`
+
+```json
+{
+    "id": "123456780",
+    "type": "https://didcomm.org/messagepickup/3.0/status-request",
+    "body" : {
+        "recipient_key": "<key for messages>"
+    },
+    "return_route": "all"
+}
+```
+`recipient_key` is optional. When specified, the `mediator` **MUST** only return status related to that recipient key. This allows the `recipient` to discover if any messages are in the queue that were sent to a specific key. You can find more details about `recipient_key` and how it's managed in [Mediator Coordination Protocol]().
```

Comment:
We should probably think how to encode the recipient_keys. The v1 and v2 of this protocol use base58 encoded keys, but this carries the assumption that all keys are ed25519 keys. We need a more expressive way to encode keys. After that we switched to `did:key` encoded keys which is also not the ideal way to express keys IMO.

What if we start using `publicKeyMultibase`? E.g. the Ed25519Signature2020 suites uses this (https://w3c-ccg.github.io/lds-ed25519-2020/). It provides the public key and also the encoding + key type. Basically it's a did:key without the did:key part. We should probably look at limiting which bases to support: https://github.com/w3c-ccg/lds-ed25519-2020/issues/3

Kyle also mentioned that we could look at JWK encoding of keys: https://github.com/hyperledger/aries-rfcs/issues/704#issuecomment-950489054. Also seems like a valid option to me, but more complex and not encodeable as a single string.

Reply:
Since DIDcomm v2 the forward messages now uses dids to forward messages. Should we start using DIDs instead of the keys? In that case we don't have to solve the encoding issues. This would also make key rotation easier and it probably makes sense to use dids in the DIDComm world.

https://identity.foundation/didcomm-messaging/spec/#messages

Reply:
> Also see my comment on the mediator coordination protocol: https://github.com/decentralized-identity/didcomm.org/pull/50/files#r923169329

yes, we should match on what we define in the mediator coordination protocol. `recipient-did` makes a lot of sense